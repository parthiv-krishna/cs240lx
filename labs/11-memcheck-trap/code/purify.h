#ifndef __PURIFY_H__
#define __PURIFY_H__

#include "src-loc.h"


void purify_init(void);
#define purify_alloc(_n) purify_alloc_raw(_n, SRC_LOC_MK())
#define purify_free(_ptr) purify_free_raw(_ptr, SRC_LOC_MK())

void *purify_alloc_raw(unsigned n, src_loc_t loc);
void purify_free_raw(void *p, src_loc_t loc);

int purify_heap_errors(void);

// just for the debug macro.
#include "memtrace.h"

static inline void purify_verbose_set(int v_p) {
    memtrace_verbose_set(v_p);
}
#endif
