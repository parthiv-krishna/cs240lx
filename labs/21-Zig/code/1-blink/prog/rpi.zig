extern fn nop() void;

// countdown 'ticks' cycles
pub fn delay(ticks: u32) void {
    var cnt = @as(u32, ticks);
    while (cnt > 0) {
        nop();
        cnt -= @as(u32, 1);
    }
}


