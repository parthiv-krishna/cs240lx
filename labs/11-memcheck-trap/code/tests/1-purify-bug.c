// checks a single memory overflow.
#include "rpi.h"
#include "purify.h"

void notmain(void) {
    trace("should detect memory overflow at 1 byte past block end\n");

    purify_init();
    char *p = purify_alloc(4);
    trace("allocated [addr=%x]: about to store\n", p);
    memset(p, 0, 4);
    p[4] = 1;   // one past end of block

    trace("should have caught the corruption before now!\n");
}
