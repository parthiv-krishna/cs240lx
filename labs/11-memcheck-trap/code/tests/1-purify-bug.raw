find-ttyusb.c:find_ttyusb:55:FOUND: </dev/ttyUSB0>
opened tty port </dev/ttyUSB0>.
pi-install: tty-usb=</dev/ttyUSB0> program=<./tests/1-purify-bug.bin> about to boot
HASH: simple_boot: sending 28756 bytes, crc32=c80b9d5a
waiting for a start
putting code
pi sent print: <<Flynn Dreilinger Lab 4>: success: Received the program!
>
bootloader: Done.
listening on ttyusb=</dev/ttyUSB0>
TRACE:should detect memory overflow at 1 byte past block end
TRACE:allocated [addr=0x183ef8]: about to store
memtrace.c:data_abort_memtrace:100:got a domain section fault at pc=0xbef4: address 0x183ef8
TRACE:[pc=0xbef4]: store to address 0x183ef8
memtrace.c:data_abort_memtrace:114:set a mismatch on pc=0xbef4 [regs[15]=0xbef4]
memtrace.c:prefetch_abort_memtrace:71:mismatch about to resume back to: 0xbef8!
memtrace.c:data_abort_memtrace:100:got a domain section fault at pc=0xbef4: address 0x183ef9
TRACE:[pc=0xbef4]: store to address 0x183ef9
memtrace.c:data_abort_memtrace:114:set a mismatch on pc=0xbef4 [regs[15]=0xbef4]
memtrace.c:prefetch_abort_memtrace:71:mismatch about to resume back to: 0xbef8!
memtrace.c:data_abort_memtrace:100:got a domain section fault at pc=0xbef4: address 0x183efa
TRACE:[pc=0xbef4]: store to address 0x183efa
memtrace.c:data_abort_memtrace:114:set a mismatch on pc=0xbef4 [regs[15]=0xbef4]
memtrace.c:prefetch_abort_memtrace:71:mismatch about to resume back to: 0xbef8!
memtrace.c:data_abort_memtrace:100:got a domain section fault at pc=0xbef4: address 0x183efb
TRACE:[pc=0xbef4]: store to address 0x183efb
memtrace.c:data_abort_memtrace:114:set a mismatch on pc=0xbef4 [regs[15]=0xbef4]
memtrace.c:prefetch_abort_memtrace:71:mismatch about to resume back to: 0xbef8!
memtrace.c:data_abort_memtrace:100:got a domain section fault at pc=0x8098: address 0x183efc
TRACE:[pc=0x8098]: store to address 0x183efc
TRACE:ERROR: illegal store to to allocated block at [pc=0x8098]: [addr=0x183efc] is 1 bytes after legal mem (block size=4)
TRACE:	logical block id=1, [addr=0x183ef8] nbytes=4
TRACE:	Block allocated at: tests/1-purify-bug.c:notmain:9
DONE!!!

Saw done

bootloader: pi exited.  cleaning up
