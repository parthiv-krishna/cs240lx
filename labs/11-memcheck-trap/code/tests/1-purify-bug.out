TRACE:should detect memory overflow at 1 byte past block end
TRACE:allocated : about to store
TRACE:: store to address 0x183ef8
TRACE:: store to address 0x183ef9
TRACE:: store to address 0x183efa
TRACE:: store to address 0x183efb
TRACE:: store to address 0x183efc
TRACE:ERROR: illegal store to to allocated block at  is 1 bytes after legal mem (block size=4)
TRACE:	logical block id=1,  nbytes=4
TRACE:	Block allocated at: tests/1-purify-bug.c:notmain:9
